src/wp-includes/class-wp-theme.php

```diff
@@ -1967,14 +1967,21 @@ public function get_block_patterns() {
 	 * Gets block pattern cache.
 	 *
 	 * @since 6.4.0
+	 * @since 6.6.0 Adds support for transient caching.
 	 *
 	 * @return array|false Returns an array of patterns if cache is found, otherwise false.
 	 */
 	private function get_pattern_cache() {
 		if ( ! $this->exists() ) {
 			return false;
 		}
-		$pattern_data = wp_cache_get( 'wp_theme_patterns_' . $this->stylesheet, 'theme_files' );
+
+		if ( wp_using_ext_object_cache() ) {
+			$pattern_data = wp_cache_get( 'wp_theme_patterns_' . $this->stylesheet, 'theme_files' );
+		} else {
+			$pattern_data = get_site_transient( 'theme_files_wp_theme_patterns_' . $this->stylesheet );
+		}
```

Comment:
Not a must have for this PR, but given that there are other places to use the `theme_files` cache group with the transient fallback, it may be worth introducing 3 private methods in this class to abstract away this `if else` and thus make it more intuitive. For example:
* `get_files_cache( $key )`
* `set_files_cache( $key, $data, $expiration )`
* `delete_files_cache( $key )`

That way the dedicated methods here could just call that, and it would make it more obvious that other theme files caching logic should reuse this. `$key` would be `'wp_theme_patterns_' . $this->stylesheet` in the code here.

Reply:
I agree with this. I think we should wait and address that once we're sure we can add caches for block templates and template parts, which I think can be handled in a separate PR.

Reply:
This if statement isn't required. 

Transients automatically use the object cache if it's available https://github.com/WordPress/wordpress-develop/blob/6456e6ea8b4ef25762a7c94821011d42a0c6da99/src/wp-includes/option.php#L2304-L2306

🔢  Applies throughout so I'll avoid repetition.

Reply:
@peterwilsoncc As part of the ticket, it was decided to use the object cache functions when a persistent cache available, as storing in a transient cache results in different behavior. So this if clause is intentional, what you're proposing would not result in the same behavior.

Reply:
@felixarntz Are you able to explain the different behavior you are worried about? I realise it would change the cache group but otherwise the timeout seems the same.

Reply:
@peterwilsoncc I don't fully recall the discussion that led to that decision (likely happened somewhere as part of https://core.trac.wordpress.org/ticket/59719), but I think the main reason is semantics and supplemental functionality like cache invalidation:
* Transients are effectively a mechanism to cache, similar to the `wp_cache_*()` functions.
* The `site-transient` cache group is effectively a "bucket" for any kind of unrelated data, only used to allow transients to be stored in persistent object cache. On the other hand, the `theme_files` cache group actually carries a meaning and groups data that belongs together.
* For instance, it makes sense to call `wp_cache_flush_group( 'theme_files' )` to clear caches once you've made theme file modifications. Calling `wp_cache_flush_group( 'site-transient' )` would be excessive and invalidate a ton of unrelated other data.
* In other words, using transients here is just a fallback for when there's no persistent object cache available. Using a dedicated cache group is more appropriate, that's why this is done when an object cache is available.

Reply:
I think the conditional makes it overly complex for site maintainers and owners.

Using a transient allows a single command be used to flush the cache regardless of the site details `wp transient delete [key] --network` or the PHP equivalent for plugins.

Requiring different approaches depending on the site internals makes it quite unintuitive.

Reply:
@peterwilsoncc the main purpose of this PR and the related ticket, is to reduce certain expensive operations related to parsing block theme template files (i.e., patterns, templates, and template parts). 

Initially, this was implemented using wp_cache_* functions and later added to a `theme_files` group to make it easier to flush this set of data when invalidation needs to be handled granularly. However, right now these caches only benefits sites that employ a persistent object cache. What we're trying to do here is extend the benefits of that persistence to all sites, while retaining the granularity that object caches already have.

Given that these caches are getting deleted any time `WP_Theme::cache_delete()` is called, maybe retaining that granularity isn't really needed? If so, then I think relying on the `site-transient` group for persistence could be fine. In practice, I'm not sure how many people are actively using the new `theme_files` group since it was just introduced.

Ideally, we would have a much friendlier way to flush a group of transients so that these theme related caches could be cleared in a consistent manner regardless of what system was being used to handle the caches.

Reply:
> To me the consistency of having it always as a transient doesn't justify the lack of granular cache control a dedicated cache group provides.

I'm not sure it provides the advantage though.

Both Core contributors and plugin developers will need to account for both situations when clearing the cache. Using a forked approach risks situations in which the cache is dumped for some systems and not others. Intermittent bugs like that are difficult to diagnose at the best of times, when it's due to system dependencies outside the developers control, it's even more difficult.

I used the WP CLI approach as a low level example: in reality maintainers and extenders will use th PHP functions `delete_site_transient` and `wp_cache_delete()`, etc to manage the caching. This will happen no matter how well documented it is to use the method in this class.

Reply:
After a detailed discussion on [slack](https://wordpress.slack.com/archives/C02KGN5K076/p1712676908690929) over whether to use a forked method or just rely on the transient function, we have concluded that transient functions wp_cache_*() provide all the benefits of a normal cache function, with the exception that we can't clear the cache by the "theme_files" group.

For those who wish to clear the cache based on a group, they can do so by the using "site-transient" group. This does have a downside in that it would clear more than just the theme file cache, but it would still not affect "high frequency" caches like posts, etc.


I have modified the code in favour of this discussion.